Rework capturing system and consume RA engineers on capture by default

pchote opened the pull request:
This PR merges the internal and external capturing logic and improves on both.  It is now possible to mix and match the following behaviours for all capturing types:
* Sabotage on/off
* External capture delay on/off
* Run inside and be consumed on/off
* Capture progress bars / blinking (via their own traits)

This should also fix the various edge cases caused by having multiple `Captures`/`Capturable` traits on a single actor, and with multiple actors capturing an actor at the same time.

This also unifies the campaign and default MP capturing logic in RA by switching engineers to use C&C3:KW-style capturing: engineers will wait outside the structure for 5 seconds, then run inside and are consumed. A "Reusable Engineers" lobby option (default off) is provided to restore the old behavior in MP.

The default behaviour change shouldn't be too much of a surprise, as I have raised this (without any real controversy) several times over the last couple of years in the discussions into the engineer/mechanic/hijacker situation, differences from the original C&Cs, and the campaign (e.g. #12806 / #14339).

My main concerns for potential regressions are:
* Campaign missions
* Disabling behaviours while a capture is underway
  * Note that #15045 will regress the behaviour for blocking production, but #15058 (and then switching the pause from `Production` to the `Exits`) fixes it properly.
  * #14632 is now also a problem for RA, but needs to be fixed in the production/exit code, not here.

This plus #15045, and #15058 together remove our last uses of the building lock, so that will shortly be removable.

There are also plans to make this change in TD, but i'm deferring that to a future PR so that the sabotage logic can be tested here.

Fixes #11493
Fixes #12474
Fixes #12400
Fixes #15076
Closes #15123
Fixes #15330
Partially fixes (now 2/3) #12604
Helps with #12806

Moving forward, I think it is reasonable to WONTFIX the following:
Closes #4875
Closes #4870
Closes #5101
Closes #14261
Closes #12979

pchote committed:
Add CaptureManager trait to fix multiple-trait interactions.

This fixes the various edge cases that occur when multiple
Captures or Capturable traits are defined on an actor and
are toggled using conditions.

The Sabotage threshold field moves from Capturable to
Captures in order to simplify the plumbing. The previous
behaviour ingame can be restored by creating a new
capturable type for each threshold level, each with their
own Captures trait.

pchote committed:
Disable sabotaging by default.

pchote committed:
Add IMove.EstimatedMoveDuration.

pchote committed:
Add ConsumedByCapture flag to emulate legacy external behaviour.

pchote committed:
Remove legacy building lock calls.

This has been superseded by conditions.

pchote committed:
Remove default building capture type.

pchote committed:
Add PreventsAutoTarget.

pchote committed:
Rework RA Engineer behaviour.

Capturing now behaves as in C&C3:KW - wait for 8 seconds
outside the structure before running inside and being disposed.

A "Reusable Engineer" lobby option is provided to restore
the previous non-disposing behaviour.

pchote committed:
Remove obsolete code.

pchote committed:
Fix captor being disposed if the capture fails.

MustaphaTR commented:
Balance can be further adjusted when we get this merged, but in my opinion. RA mod should keep Generals Style the long timer reusable engineers. Afaik ZxGanon has plans to introduce the KW like short delay consumed engis to TD mod. D2k can keep the old behaviur so we have all the engi logics shown in our main mods. Best probably would be to hear more from the community tho.

Also about the blnking trait? Can we get ability to play a sound for each blink and make delay between blinks get shorter and shorter as capture goes on as in Generals? Would be nice to be able to replicate behaviour there.

ZxGanon commented:
Yes I have been waiting for so long for this change.

I can´t wait to implement this into Shattered Paradise since this is just the most modern and balanced way to implement engineer capturing.

Also since lobby options are impelmented everyone can customize it (if not implementing it as a must have like we are gonna do in SP).

pchote commented:
Bumped the capture delay to 200 ticks (8 seconds) as discussed in IRC.

pchote committed:
Remove obsolete building lock check from BaseProvider.

pchote commented:
Updated:
 * Rebased on top of #15643
 * Added a new commit to remove the now unused building lock check from `BaseProvider`

I also confirmed that #15650 does not occur with the new logic.

pchote committed:
Add capture progress bars and blinking.

abcdefg30 commented on OpenRA.Mods.Common/Traits/CaptureManager.cs:
```diff
@@ -0,0 +1,278 @@
+
+			if (conditionManager != null && !string.IsNullOrEmpty(info.CapturingCondition) &&
+					capturingToken == ConditionManager.InvalidConditionToken)
+				capturingToken = conditionManager.GrantCondition(self, info.CapturingCondition);
```
Why do we grant the condition even if `captures == null` below (and the method returns `false`)?

pchote replied:
From the method description
```csharp
/// Called by CaptureActor when the activity is ready to enter and capture the target.
/// This method grants the capturing conditions on the captor and target and returns
/// true if the captor is able to start entering or false if it needs to wait.
```
The conditions are active for the whole capture period (including the delay).  If we return false because the target is no longer capturable, then the code from `CaptureActor` will kick in to cancel the activity, and the condition will be disabled at the appropriate time.

pchote replied:
Renamed `Enter.CanEnter` to `Enter.TryStartEnter` and `CaptureManager.TryCapture` to `CaptureManager.StartCapture` to hopefully fix this confusion.

abcdefg30 commented on mods/d2k/rules/infantry.yaml:
```diff
@@ -41,8 +41,9 @@ engineer:
+	CaptureManager:
 	Captures:
-		CaptureTypes: building, husk
+		CaptureTypes: building
```
This probably wants to be reverted again (as discussed on IRC).

pchote replied:
As discussed in IRC, this was leftover from #9615 which removed the capturing.

pchote committed:
Add support for capture delays and conditions.

CaptureDelay is defined per-trait, and the shortest
delay will be used if multiple traits are enabled.

CapturingCondition and BeingCapturedCondition are
global, and are granted when any capture is in progress.

The capture period is defined from when the unit reaches
the cell next to the target, and either starts to wait
or enter the target through to when the capture succeeds
or fails.

pchote committed:
Add MergeCaptureTraits update rule.

pchote commented:
Fixed.

pchote committed:
Stop vehicle movement when a hijacker is entering.

pchote commented:
Updated:
 * Added a commit that uses `BeingCapturedCondition` to stop target movement when a hijacker captures a vehicle.  This should make them much more usable.

pchote committed:
Remove unwanted overrides from allies-03*.

abcdefg30 reviewed (approved):
Looks good to me now. Great work! :)

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/dbe32ce6589bf6086d78e9ac15eff68d4bec94d9...50013139f4d559b6950fe204f00612a7b7314403)

Punsho commented:
Will the engineer lobby option also toggle the ability to block the production, disallow selling? If not this seems like just a downgrade from the gameplay standpoint

pchote commented:
> Will the engineer lobby option also toggle the ability to block the production, disallow selling? If not this seems like just a downgrade from the gameplay standpoint

No, these remain disabled in both modes. Enabling selling during capture seems a bit risky IMO, but we can consider toggling this if there is enough demand during the playtests.

pchote commented:
The 8 second capture delay has been heavily panned by the discord community.  We'll want to drop this back, perhaps even down to 3 or 4 sec.
